### PR TITLE
Layers should inherit group z-index

### LINKED ISSUE #6654: Documentation for zIndex in regards to ol.layer.Group needs clarification
It is unclear in the documentation as to how `zIndex` is handled when rendering `ol.layer.Group` objects with sub layers.

After digging into the source code a little, it appears to me that [`zIndex` set on the parent layer (`ol.layer.Group`) will not be inherited by the child layer](https://github.com/openlayers/openlayers/blob/ac84e560bc4a2b7e0e37cf306decf0e676a9cf05/src/ol/layer/group.js#L208).  So if no `zIndex` is set on the child layers, their `zIndex` will default to zero (even if the parent layer has a `zIndex` set to 100) and any non-group layer that is added to the map with a non-zero `zIndex` will be rendered in front of them.  Is `zIndex` on `ol.layer.Group` ever respected in the rendering context?  If so can we clarify how it interacts with the child layers and other layers on the map?

https://jsfiddle.net/wajxb9eb/1/

### LINKED ISSUE #12059: LayerGroup zIndex is not inherited by the child layers
**Describe the bug**
I noticed that the zIndex of the LayerGroup is not inherited by the child layer added to it. Is this expected behaviour or a bug? Reason for this question is that i found an old bug report from years back of the same thing and there was a comment that it should be inherited. There was also PR made, but never merged?

https://github.com/openlayers/openlayers/issues/6654

**To Reproduce**
Example code can be provided if needed.

**Expected behavior**
Based on the LayerGroup documentation, zIndex of the group should be inherited by the child layers.

### PR DESCRIPTION
**Pull request by MoonE**
This will let layers in a layer group inherit the group's zIndex if no own zIndex is defined.

Somewhat based on #6661
Fixes #6654, fixes #12059

### CODE REVIEW

File: test/browser/spec/ol/layer/group.test.js [hunk 1]
```diff
@@ -497,6 +501,67 @@ describe('ol.layer.Group', function () {
...
+        ],
+      });
+
+      const layerStatesArray = group.getLayerStatesArray();
+      expect(layerStatesArray[0].zIndex).to.be(0);
+
+      disposeHierarchy(group);
+    });
+
+    it('returns undefined zindex in call to getLayerStatesArray when nested', function () {
```
Comment (tschaub) [inline test/browser/spec/ol/layer/group.test.js:552]: This description (“when nested”) looks more appropriate for the case above.

File: src/ol/layer/Group.js [hunk 2]
```diff
@@ -205,19 +205,27 @@ class LayerGroup extends BaseLayer {
...
     const pos = states.length;
 
     this.getLayers().forEach(function (layer) {
       layer.getLayerStatesArray(states);
     });
 
     const ownLayerState = this.getLayerState();
+    let defaultZIndex = ownLayerState.zIndex;
+    if (!opt_states && ownLayerState.zIndex === undefined) {
+      defaultZIndex = 0;
```
Comment (tschaub) [inline src/ol/layer/Group.js:227]: It isn’t clear to me why this is necessary to solve the problem.

File: test/browser/spec/ol/layer/group.test.js (hunk 1 above)
Comment (MoonE) [reply test/browser/spec/ol/layer/group.test.js]: nested in this case referred to the `getLayerStatesArray([])` call with passed states array which indicates that it is called from a parent layer group..

I updated the tests, this was indeed confusing.

File: src/ol/layer/Group.js (hunk 2 above)
Comment (tschaub) [reply src/ol/layer/Group.js:227]: > At the end of the `getLayerStatesArray` call chain all layer states should have a defined zIndex

Kind of subtle that `getLayerState` will return `zIndex: undefined` and `getLayerStates` will return `zIndex: 0`.  I don't have my head enough into this to know if this is what we want or not.  Just that it tripped me up when reviewing the test changes.

**Review by tschaub — APPROVED** [review]

**Comment by ahocevar** [comment]
Did you mean to merge this, @MoonE?

**Comment by MoonE** [comment]
I do, thanks for the reminder.

**Pull request merged**